libflux/src/core/parser/mod.rs

```diff
@@ -559,7 +559,29 @@ impl Parser {
         // Tvar | Basic | Array | Record | Function
         let t = self.peek();
         match t.tok {
-            TOK_LBRACK => self.parse_array_type(),
+            TOK_LBRACK => {
+                let start = self.open(TOK_LBRACK, TOK_RBRACK);
+                let ty = self.parse_monotype();
+                match self.peek().tok {
+                    TOK_RBRACK => {
+                        let end = self.close(TOK_RBRACK);
+                        MonoType::Array(Box::new(ArrayType {
+                            base: self.base_node_from_tokens(&start, &end),
+                            element: ty,
+                        }))
+                    }
+                    _ => {
+                        self.expect(TOK_COLON);
+                        let val = self.parse_monotype();
+                        let end = self.close(TOK_RBRACK);
+                        MonoType::Dict(Box::new(DictType {
+                            base: self.base_node_from_tokens(&start, &end),
+                            key: ty,
+                            val: val,
+                        }))
+                    }
+                }
+            }
```

Comment:
This is the block of code that makes it so that we can parse dict types in builtin statements.